eth/eventdatahandler/event_data_handler.go

```diff
@@ -0,0 +1,525 @@
+
+		if !updated {
+			return nil, nil
+		}
+
+		task := NewUpdateFeeRecipientTask(edh.taskExecutor, feeRecipientAddressUpdatedEvent.Owner, feeRecipientAddressUpdatedEvent.RecipientAddress)
+		return task, nil
+
+	default:
+		edh.logger.Warn("unknown event name", fields.Name(abiEvent.Name))
+		return nil, nil
+	}
+}
+
+func (edh *EventDataHandler) HandleLocalEvents(localEvents []localevents.Event) error {
+	txn := edh.eventDB.RWTxn()
+	defer txn.Discard()
+
+	for _, event := range localEvents {
+		if err := edh.processLocalEvent(txn, event); err != nil {
+			return fmt.Errorf("process local event: %w", err)
+		}
+	}
+
+	if err := txn.Commit(); err != nil {
+		return fmt.Errorf("commit transaction: %w", err)
+	}
+
+	return nil
+}
+
+func (edh *EventDataHandler) processLocalEvent(txn eventdb.RW, event localevents.Event) error {
+	switch event.Name {
+	case OperatorAdded:
+		data := event.Data.(contract.ContractOperatorAdded)
+		if err := edh.handleOperatorAdded(txn, &data); err != nil {
+			return fmt.Errorf("handle OperatorAdded: %w", err)
+		}
+		return nil
+	case OperatorRemoved:
+		data := event.Data.(contract.ContractOperatorRemoved)
+		if err := edh.handleOperatorRemoved(txn, &data); err != nil {
+			return fmt.Errorf("handle OperatorRemoved: %w", err)
+		}
+		return nil
+	case ValidatorAdded:
+		data := event.Data.(contract.ContractValidatorAdded)
+		if _, err := edh.handleValidatorAdded(txn, &data); err != nil {
+			return fmt.Errorf("handle ValidatorAdded: %w", err)
+		}
+		return nil
+	case ValidatorRemoved:
+		data := event.Data.(contract.ContractValidatorRemoved)
+		if _, err := edh.handleValidatorRemoved(txn, &data); err != nil {
+			return fmt.Errorf("handle ValidatorRemoved: %w", err)
+		}
+		return nil
+	case ClusterLiquidated:
+		data := event.Data.(contract.ContractClusterLiquidated)
+		_, err := edh.handleClusterLiquidated(txn, &data)
+		if err != nil {
+			return fmt.Errorf("handle ClusterLiquidated: %w", err)
+		}
+		return nil
+	case ClusterReactivated:
+		data := event.Data.(contract.ContractClusterReactivated)
+		_, err := edh.handleClusterReactivated(txn, &data)
+		if err != nil {
+			return fmt.Errorf("handle ClusterReactivated: %w", err)
+		}
+		return nil
+	case FeeRecipientAddressUpdated:
+		data := event.Data.(contract.ContractFeeRecipientAddressUpdated)
+		_, err := edh.handleFeeRecipientAddressUpdated(txn, &data)
+		if err != nil {
+			return fmt.Errorf("handle FeeRecipientAddressUpdated: %w", err)
+		}
+		return nil
+	default:
+		edh.logger.Warn("unknown local event name", fields.Name(event.Name))
+		return nil
+	}
+}
+
+// filterSupersedingTasks filters out tasks that are superseded by other tasks:
+// - opposite tasks (start-stop, stop-start, liquidate-reactivate, reactivate-liquidate)
+// - superseding tasks (updateFee-updateFee)
+// TODO: check if start-stop can be superseded by reactivate-liquidate
+func (edh *EventDataHandler) filterSupersedingTasks(tasks []Task) []Task {
```

Comment:
I must say this whole process kinda scares me, if we end up removing some task that needs to happen or not removing one that shouldn't we risk being slashed. for example aren't there are more consequences to removing a pair of start-stop? for example other events in between that might happen ( updateFee?) what should we do with these.
I understand why we do this I just think we need to thoroughly make sure we do the right thing and have tests in place.

Reply:
I made it optional